tests/api/v2/test_knowledge.py

```diff
@@ -176,6 +181,51 @@ async def test_add_fact(knowledge_webapp, aiohttp_client):
     assert current == response
 
 
+async def test_add_fact_to_operation(knowledge_webapp, aiohttp_client, test_operation, setup_empty_operation):
+    client = await aiohttp_client(knowledge_webapp)
+
+    fact_data = {
+        'trait': 'demo',
+        'value': 'test',
+        'source': test_operation['id']
+    }
+    resp = await client.post('/facts', json=fact_data, headers=headers)
+    data = await resp.json()
+    response = data['added']
+    assert len(response) == 1
+    assert response[0]['trait'] == 'demo'
+    assert response[0]['value'] == 'test'
+    assert response[0]['source'] == test_operation['id']
+
+    tmp = await client.get('/facts', json=fact_data, headers=headers)
+    cur = await tmp.json()
+    current = cur['found']
+    assert current == response
+    data_svc = BaseService.get_service('data_svc')
+    file_svc = BaseService.get_service('file_svc')
+    matched_operations = await data_svc.locate('operations', {'id': test_operation['id']})
+    report = await matched_operations[0].report(file_svc, data_svc)
+    assert response[0] in report['facts']
+
+
+async def test_add_fact_to_finished_operation(knowledge_webapp, aiohttp_client, setup_finished_operation,
+                                              finished_operation_payload):
+    client = await aiohttp_client(knowledge_webapp)
+    op_id = finished_operation_payload['id']
+    matched_operations = await BaseService.get_service('data_svc').locate('operations', {'id': op_id})
+    assert await matched_operations[0].is_finished()
+
+    fact_data = {
+        'trait': 'demo',
+        'value': 'test',
+        'source': op_id
+    }
+    resp = await client.post('/facts', json=fact_data, headers=headers)
+    data = await resp.json()
+    assert 'error' in data
+    assert 'Cannot add fact to finished operation.' in data['error']
```

Comment:
are you able to assert that the error class matches? (Sorry, just making sure we're thorough here)

Reply:
Sorry forgot to comment on the other thread - this exception is caught by a more general try/catch in `fact_api.py`, so the only error type specific content I can fetch is the actual error message (the general try/catch just returns a JSONBadRequest error with the exception-specific error message)

Reply:
There are ways to get the full error stack when testing, but given the intended use case for this API, I think this approach is valid enough.